In [1]:
from google.colab import drive
drive.mount('/content/drive/')

Mounted at /content/drive/


In [2]:
!pip install timm segmentation_models_pytorch rasterio torchmetrics

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.2/2.2 MB 29.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 106.7/106.7 kB 13.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 20.6/20.6 MB 86.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 805.2/805.2 kB 66.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 302.0/302.0 kB 35.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.3/1.3 MB 80.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.8/58.8 kB 8.4 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.2/2.2 MB 77.1 MB/s eta 0:00:00
  Created wheel for efficientnet-pytorch: filename=efficientnet_pytorch-0.7.1-py3-none-any.whl size=16428 sha256=cae7e42862554e10ec5139bef3bc432e192180f4b294193041db58e1b725dac6
  Stored in directory: /root/.cache/pip/wheels/03/3f/e9/911b1bc46869644912bda90a56bcf7b960f20b5187f

In [3]:
import numpy as np
import torch
from torch.utils.data import Dataset, DataLoader
from sklearn.metrics import f1_score
import torch.nn.functional as F

In [18]:
params = {
    'backbone': 'timm-efficientnet-b1',
    'weights': 'imagenet', # 'noisy-student',
    'learning_rate': 0.001,
    'epochs': 50,
    'batch_size': 30,
    'num_workers': 12,
    'device': torch.device('cuda') if torch.cuda.is_available() else torch.device('cpu'),
    'div_factor':10,
    'final_div_factor':100,
    'threshhold':0.5
}

## Functions

We define a function on which we can measure the accuracy of our model.

In [5]:
def sort_paths_by_number(paths):
    return sorted(paths, key=lambda x: int(x.split('_')[-1].split('.tif')[0]))

In [6]:
def calculate_f1_score(pred, true, threshold=0.5):
    """
    Compute the F1 score, also known as balanced F-score or F-measure.

    Parameters:
    - true: ground truth (correct) labels, a tensor of shape [N, H, W].
    - pred: predicted labels, a tensor of shape [N, 1, H, W].
    - threshold: the prediction threshold, defaults to 0.5.

    Returns:
    - F1 score
    """

    # Flatten the tensors
    true = true.numpy().flatten()
    pred = pred.numpy().flatten()

    # Calculate F1 score
    f1 = f1_score(true, pred)

    return f1

In [7]:
def intersection_and_union(pred, true):
    """
    Calculates intersection and union for a batch of images.

    Args:
        pred (torch.Tensor): a tensor of predictions
        true (torc.Tensor): a tensor of labels

    Returns:
        intersection (int): total intersection of pixels
        union (int): total union of pixels
    """
    pred = pred.squeeze(1)
    valid_pixel_mask = true.ne(255)  # valid pixel mask
    true = true.masked_select(valid_pixel_mask)
    pred = pred.masked_select(valid_pixel_mask)

    # Convert to CPU and NumPy for logical operations
    true = true.numpy()
    pred = pred.numpy()

    # Intersection and union totals
    intersection = np.logical_and(true, pred)
    union = np.logical_or(true, pred)
    return intersection.sum(), union.sum()

In [9]:
def save_model(epoch,model,ckpt_path='./',name='unet_effnet_b1_1000x1000',val_iou=0):
    path = os.path.join(ckpt_path, '{}_wo_ca.pth'.format(name))
    torch.save(model.state_dict(), path, _use_new_zipfile_serialization=False)

def load_model(model,ckpt_path):
    state = torch.load(ckpt_path)
    model.load_state_dict(state)
    return model

## Loss Functions

In [10]:
class FocalLoss(torch.nn.modules.loss._WeightedLoss):
    def __init__(self, weight=None, gamma=2,reduction='mean'):
        super(FocalLoss, self).__init__(weight,reduction=reduction)
        self.gamma = gamma
        self.weight = weight #weight parameter will act as the alpha parameter to balance class weights
        self.bce = torch.nn.BCEWithLogitsLoss(reduction="none")

    def forward(self, input, target):
        #ce_loss = F.binary_cross_entropy_with_logits(input.squeeze(1), target.long(),reduction=self.reduction,weight=self.weight)
        ce_loss = self.bce(input.squeeze(1), target.squeeze(1))
        pt = torch.exp(-ce_loss)
        focal_loss = ((1 - pt) ** self.gamma * ce_loss).mean()
        return focal_loss

In [11]:
class XEDiceLoss(torch.nn.Module):
    """
    Computes (0.5 * CrossEntropyLoss) + (0.5 * DiceLoss).
    """

    def __init__(self):
        super().__init__()
        self.xe = torch.nn.BCEWithLogitsLoss(reduction="none")
        #self.xe = FocalLoss(reduction='none')

    def forward(self, pred, true):

        pred = pred.squeeze(1)
        valid_pixel_mask = true.ne(255)  # valid pixel mask

        # Cross-entropy loss
        temp_true = torch.where((true == 255), 0, true)  # cast 255 to 0 temporarily
        xe_loss = self.xe(pred, temp_true.float())
        xe_loss = xe_loss.masked_select(valid_pixel_mask).mean()

        # Dice loss

        pred = pred.sigmoid() #torch.softmax(pred, dim=1)[:, 1]

        pred = pred.masked_select(valid_pixel_mask)
        true = true.masked_select(valid_pixel_mask)

        dice_loss = 1 - (2.0 * torch.sum(pred * true) + 1e-7) / (torch.sum(pred + true) + 1e-7)

        #print(xe_loss,dice_loss,(2.0 * torch.sum(pred * true) + 1e-7))

        return (0.5 * xe_loss) + (0.5 * dice_loss)

## Data
We create the Dataset and Dataloader

In [12]:
import albumentations as A
from albumentations.pytorch import ToTensorV2
from sklearn.model_selection import train_test_split

import glob
import pandas as pd
import tifffile
import os

In [13]:
PATH_SRC = f'/content/drive/MyDrive/Projekte/Solafune/Cloud_satelite/src'
PATH_OUTPUT = f'/content/drive/MyDrive/Projekte/Solafune/Cloud_satelite/submit'
PATHS_VAL = sort_paths_by_number(glob.glob(f'/content/drive/MyDrive/Projekte/Solafune/Cloud_satelite/evaluation_true_color/evaluation_true_color_*.tif'))
PATHS_VAL_MASK = sort_paths_by_number(glob.glob(f'/content/drive/MyDrive/Projekte/Solafune/Cloud_satelite/sample/evaluation_mask_*.tif'))
PATHS_TRAIN = sort_paths_by_number(glob.glob(f'/content/drive/MyDrive/Projekte/Solafune/Cloud_satelite/train_true_color/train_true_color_*.tif'))
PATHS_TRAIN_MASK = sort_paths_by_number(glob.glob(f'/content/drive/MyDrive/Projekte/Solafune/Cloud_satelite/train_mask/train_mask_*.tif'))

In [17]:
class DataTransform():
    def __init__(self):
        self.data_transform = {
            "train": A.Compose(
                [
                  A.ShiftScaleRotate(shift_limit=0.0625,rotate_limit=15,p=0.5),
                  A.GridDistortion(p=0.35),
                  A.HorizontalFlip(p=0.5),
                  A.VerticalFlip(p=0.5),
                  A.GaussianBlur(p=0.25),
                  A.PadIfNeeded(min_height=1024, min_width=1024, border_mode=0),
                  # A.Resize(512, 512, interpolation=1, p=1),
                  A.RandomCrop(384,384),
                  ToTensorV2()
                ]
            ),
            "val": A.Compose(
                [
                  A.PadIfNeeded(min_height=1024, min_width=1024, border_mode=0),
                  # A.Resize(512, 512, interpolation=1, p=1),
                  ToTensorV2()
                ]
            )
        }

    def __call__(self, phase, img, mask):
      return self.data_transform[phase](image=img, mask=mask)

In [19]:
class CloudDataset(Dataset):
  def __init__(self, df, phase, transform):
    self.data = df
    self.phase = phase
    self.transform = transform

  def __len__(self):
    return len(self.data)

  def normalize_data(self, image_tiff):
    # Normalize the image for better visualization
    for i in range(3):  # Assuming the image has 3 channels
        image_tiff[:,:,i] = (image_tiff[:,:,i] - image_tiff[:,:,i].min()) / (image_tiff[:,:,i].max() - image_tiff[:,:,i].min())

    return image_tiff

  def __getitem__(self, index):
    row = self.data.iloc[index]
    true_color_PATH = row['true_color']
    mask_PATH = row['mask']
    img = tifffile.imread(true_color_PATH)
    img = self.normalize_data(img.astype(np.float32))
    mask = tifffile.imread(mask_PATH)
    mask = mask.astype(np.float32)

    img_and_mask = self.transform(self.phase, img, mask)
    img = img_and_mask["image"]
    mask = img_and_mask["mask"]

    return img, mask

In [20]:
df = pd.DataFrame({
    'true_color': PATHS_TRAIN,
    'mask': PATHS_TRAIN_MASK
})

pd.set_option('display.max_columns', None)
pd.set_option('display.expand_frame_repr', False)
pd.set_option('max_colwidth', -1)
print(df.head())

df_train, df_val = train_test_split(df, test_size=0.2, random_state=12)

train_dataset = CloudDataset(df_train, phase="train", transform=DataTransform())
train_dataloader = DataLoader(
    train_dataset,
    batch_size=params['batch_size'],
    shuffle=True,
    num_workers=params['num_workers'],
    pin_memory=True,
    persistent_workers=True
)


val_dataset = CloudDataset(df_val, phase="val", transform=DataTransform())
val_dataloader = DataLoader(
    val_dataset,
    batch_size=2,#params['batch_size'],
    shuffle=True,
    num_workers=params['num_workers'],
    pin_memory=True,
    persistent_workers=True
)

dataloaders_dict = {
    'train': train_dataloader,
    'val': val_dataloader
}

                                                                                        true_color                                                                                 mask
0  /content/drive/MyDrive/Projekte/Solafune/Cloud_satelite/train_true_color/train_true_color_0.tif  /content/drive/MyDrive/Projekte/Solafune/Cloud_satelite/train_mask/train_mask_0.tif
1  /content/drive/MyDrive/Projekte/Solafune/Cloud_satelite/train_true_color/train_true_color_1.tif  /content/drive/MyDrive/Projekte/Solafune/Cloud_satelite/train_mask/train_mask_1.tif
2  /content/drive/MyDrive/Projekte/Solafune/Cloud_satelite/train_true_color/train_true_color_2.tif  /content/drive/MyDrive/Projekte/Solafune/Cloud_satelite/train_mask/train_mask_2.tif
3  /content/drive/MyDrive/Projekte/Solafune/Cloud_satelite/train_true_color/train_true_color_3.tif  /content/drive/MyDrive/Projekte/Solafune/Cloud_satelite/train_mask/train_mask_3.tif
4  /content/drive/MyDrive/Projekte/Solafune/Cloud_satelite/train_true_color/trai

<ipython-input-20-15fac3f26a46>:8: FutureWarning: Passing a negative integer is deprecated in version 1.0 and will not be supported in future version. Instead, use None to not limit the column width.
  pd.set_option('max_colwidth', -1)


## Model
We implement the cloud detection Model based on the Microsoft Competion Winner: https://github.com/drivendataorg/cloud-cover-winners/blob/main/1st%20place/src/unet_efficient_net_b1_train_fp16.py

In [21]:
from timm.models.efficientnet import *
import segmentation_models_pytorch as smp
import torch.nn as nn
import yaml

In [22]:
class Net(nn.Module):
  def __init__(self,params):
    super(Net, self).__init__()

    aux_params=dict(
        pooling = 'avg',             # one of 'avg', 'max'
        dropout = 0.3,               # dropout ratio, default is None
        activation = None,           # activation function, default is None
        classes = 1,
    )

    self.unet = smp.Unet(
        encoder_name = params['backbone'],        # choose encoder, e.g. mobilenet_v2 or efficientnet-b7
        encoder_weights = params['weights'],      # use `imagenet` pre-trained weights for encoder initialization
        in_channels = 3,                          # model input channels (1 for gray-scale images, 3 for RGB, etc.)
        decoder_attention_type = None,            # model output channels (number of classes in your dataset)
        classes = 1,
        aux_params = aux_params
    )

  def forward(self, image):
    batch_size = len(image)
    mask,logit = self.unet(image)
    return mask


In [23]:
model = Net(params)
model.to(params['device'])

Downloading: "https://github.com/huggingface/pytorch-image-models/releases/download/v0.1-weights/tf_efficientnet_b1-5c1377c4.pth" to /root/.cache/torch/hub/checkpoints/tf_efficientnet_b1-5c1377c4.pth
100%|██████████| 30.1M/30.1M [00:02<00:00, 10.9MB/s]


Net(
  (unet): Unet(
    (encoder): EfficientNetEncoder(
      (conv_stem): Conv2d(3, 32, kernel_size=(3, 3), stride=(2, 2), padding=(1, 1), bias=False)
      (bn1): BatchNormAct2d(
        32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True
        (drop): Identity()
        (act): Swish()
      )
      (blocks): Sequential(
        (0): Sequential(
          (0): DepthwiseSeparableConv(
            (conv_dw): Conv2d(32, 32, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1), groups=32, bias=False)
            (bn1): BatchNormAct2d(
              32, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True
              (drop): Identity()
              (act): Swish()
            )
            (se): SqueezeExcite(
              (conv_reduce): Conv2d(32, 8, kernel_size=(1, 1), stride=(1, 1))
              (act1): Swish()
              (conv_expand): Conv2d(8, 32, kernel_size=(1, 1), stride=(1, 1))
              (gate): Sigmoid()
            )
            (conv_pw): C

## Training

In [24]:
from tqdm import tqdm
import matplotlib.pyplot as plt

In [25]:
optimizer = torch.optim.Adam(model.parameters(), lr=params['learning_rate'])

scheduler = torch.optim.lr_scheduler.OneCycleLR(
      optimizer,
      steps_per_epoch = 42,
      pct_start = 0.1,
      max_lr = params['learning_rate'],
      epochs = params['epochs'],
      div_factor = params['div_factor'],
      final_div_factor = params['final_div_factor'],
      verbose = True
    )

criterion = XEDiceLoss()

Adjusting learning rate of group 0 to 1.0000e-04.


In [26]:
def train(model, img, mask):
  model.train()
  optimizer.zero_grad()

  preds = model(img)
  loss = criterion(preds, mask)
  # loss = criterion(preds.squeeze(dim=1), mask.squeeze(dim=1))
  loss.backward()
  optimizer.step()
  loss = loss.item()

  # Calculate validation IOU (global)
  preds = (preds.detach().sigmoid().cpu() > params['threshhold']) *1
  mask = mask.detach().cpu()

  # plt.imshow(preds[0][0], cmap='gray')
  # plt.colorbar()
  # plt.title("pred")
  # plt.show()

  # plt.imshow(mask[0], cmap='gray')
  # plt.colorbar()
  # plt.title("mask")
  # plt.show()


  intersection, union = intersection_and_union(preds, mask)
  f1_score = calculate_f1_score(preds, mask)

  return loss, intersection, union, f1_score

In [27]:
def val(model, img, mask):
  model.eval()
  with torch.no_grad():
    preds = model(img)
    loss = criterion(preds, mask)
    # loss = criterion(preds.squeeze(dim=1), mask.squeeze(dim=1))
    loss = loss.item()

    preds = (preds.detach().sigmoid().cpu() > params['threshhold']) *1 #torch.softmax(preds, dim=1)[:, 1]
    mask = mask.detach().cpu()
    intersection, union = intersection_and_union(preds, mask)
    f1_score = calculate_f1_score(preds, mask)

  return loss, intersection, union, f1_score

Delete GPU-RAM

In [ ]:
import gc  # Garbage Collector
gc.collect()
torch.cuda.empty_cache()

In [28]:
best_iou = 0

for epoch in range(1, params['epochs']+1):
    print(f'\nEpoch: {epoch}')
    for phase in ['train', 'val']:
        total_intersection=0
        total_union=0
        total_loss=0
        total_f1=0

        pbar = tqdm(enumerate(dataloaders_dict[phase]),
                              total=len(dataloaders_dict[phase]),
                              leave=True,
                              dynamic_ncols=True)
        for i, (img, mask) in pbar:
            img, mask = img.to(params['device'], non_blocking=True), mask.to(params['device'], non_blocking=True)

            # training
            if phase == 'train':
                loss, intersection, union, f1 = train(model, img, mask)
            # validation
            else:
                loss, intersection, union, f1 = val(model, img, mask)

            # free GPU
            del img
            del mask

            total_intersection += intersection
            total_union += union
            total_loss += loss
            iou = total_intersection/total_union
            total_f1 += f1
            pbar.set_postfix({'iou': iou, 'loss': total_loss/(i+1), 'f1': total_f1/(i+1)})

        total_loss /= len(dataloaders_dict[phase])
        total_f1 /= len(dataloaders_dict[phase])

        if phase == 'train':
          scheduler.step()

        print(f'    Phase: {phase} -->  loss={total_loss} iou={iou} f1={total_f1}')

    # save best model
    if iou > best_iou:
      print(f'Saving model with iou {iou}')
      save_model(epoch, model, ckpt_path=PATH_SRC, val_iou=iou)
      best_iou = iou
    else:
      print(f'Not saving for iou {iou}')




Epoch: 1


100%|██████████| 27/27 [03:02<00:00,  6.77s/it, iou=0.357, loss=0.572, f1=0.526]

Adjusting learning rate of group 0 to 1.0005e-04.
    Phase: train -->  loss=0.5716550151507059 iou=0.35670597162592765 f1=0.5258850673317558



100%|██████████| 100/100 [01:54<00:00,  1.14s/it, iou=0.509, loss=0.569, f1=0.53]


    Phase: val -->  loss=0.568798825442791 iou=0.50897077259596 f1=0.5298050798763958
Saving model with iou 0.50897077259596

Epoch: 2


100%|██████████| 27/27 [01:23<00:00,  3.09s/it, iou=0.572, loss=0.448, f1=0.723]


Adjusting learning rate of group 0 to 1.0020e-04.
    Phase: train -->  loss=0.44838825751233985 iou=0.5718709009618658 f1=0.7231733740030436


100%|██████████| 100/100 [01:44<00:00,  1.05s/it, iou=0.631, loss=0.451, f1=0.631]


    Phase: val -->  loss=0.4509770756959915 iou=0.631089093811803 f1=0.631135862137809
Saving model with iou 0.631089093811803

Epoch: 3


100%|██████████| 27/27 [01:22<00:00,  3.04s/it, iou=0.679, loss=0.372, f1=0.803]


Adjusting learning rate of group 0 to 1.0046e-04.
    Phase: train -->  loss=0.3716386048882096 iou=0.6786016169697869 f1=0.8031711261472299


100%|██████████| 100/100 [01:43<00:00,  1.04s/it, iou=0.714, loss=0.4, f1=0.656]


    Phase: val -->  loss=0.40024746760725977 iou=0.7137319606728701 f1=0.656321672277697
Saving model with iou 0.7137319606728701

Epoch: 4


100%|██████████| 27/27 [01:22<00:00,  3.05s/it, iou=0.718, loss=0.325, f1=0.818]


Adjusting learning rate of group 0 to 1.0081e-04.
    Phase: train -->  loss=0.3251226511266496 iou=0.7177063022924445 f1=0.818336505073134


100%|██████████| 100/100 [01:47<00:00,  1.07s/it, iou=0.738, loss=0.372, f1=0.679]


    Phase: val -->  loss=0.37192669697105885 iou=0.7383641992138099 f1=0.678770967858024
Saving model with iou 0.7383641992138099

Epoch: 5


100%|██████████| 27/27 [01:22<00:00,  3.06s/it, iou=0.735, loss=0.286, f1=0.844]


Adjusting learning rate of group 0 to 1.0127e-04.
    Phase: train -->  loss=0.2860177954037984 iou=0.735340378448882 f1=0.8438368473593988


100%|██████████| 100/100 [01:43<00:00,  1.04s/it, iou=0.757, loss=0.347, f1=0.685]


    Phase: val -->  loss=0.3472147663682699 iou=0.7574492074106006 f1=0.6852208584760033
Saving model with iou 0.7574492074106006

Epoch: 6


100%|██████████| 27/27 [01:21<00:00,  3.01s/it, iou=0.737, loss=0.268, f1=0.839]


Adjusting learning rate of group 0 to 1.0183e-04.
    Phase: train -->  loss=0.26754261387719047 iou=0.7373657684774654 f1=0.8387694249640971


100%|██████████| 100/100 [01:45<00:00,  1.05s/it, iou=0.767, loss=0.316, f1=0.721]


    Phase: val -->  loss=0.316265459805727 iou=0.7665905035642707 f1=0.7212593580673794
Saving model with iou 0.7665905035642707

Epoch: 7


100%|██████████| 27/27 [01:22<00:00,  3.05s/it, iou=0.739, loss=0.254, f1=0.839]


Adjusting learning rate of group 0 to 1.0249e-04.
    Phase: train -->  loss=0.2538286723472454 iou=0.7393523487831413 f1=0.8392874704111299


100%|██████████| 100/100 [01:46<00:00,  1.07s/it, iou=0.78, loss=0.302, f1=0.735]


    Phase: val -->  loss=0.30155374683439734 iou=0.7804349109453357 f1=0.7352886671167833
Saving model with iou 0.7804349109453357

Epoch: 8


100%|██████████| 27/27 [01:20<00:00,  2.99s/it, iou=0.756, loss=0.233, f1=0.855]


Adjusting learning rate of group 0 to 1.0325e-04.
    Phase: train -->  loss=0.23316836798632587 iou=0.756189760831393 f1=0.8551335539647608


100%|██████████| 100/100 [01:42<00:00,  1.02s/it, iou=0.773, loss=0.301, f1=0.727]


    Phase: val -->  loss=0.3007986141368747 iou=0.7726726489579667 f1=0.7268573226077109
Not saving for iou 0.7726726489579667

Epoch: 9


100%|██████████| 27/27 [01:22<00:00,  3.05s/it, iou=0.794, loss=0.208, f1=0.867]


Adjusting learning rate of group 0 to 1.0411e-04.
    Phase: train -->  loss=0.20791970634901966 iou=0.7936337047229903 f1=0.8673152980723162


100%|██████████| 100/100 [01:45<00:00,  1.05s/it, iou=0.776, loss=0.287, f1=0.728]


    Phase: val -->  loss=0.2874412129819393 iou=0.7762747472601025 f1=0.7275183044055776
Not saving for iou 0.7762747472601025

Epoch: 10


100%|██████████| 27/27 [01:22<00:00,  3.05s/it, iou=0.767, loss=0.211, f1=0.854]


Adjusting learning rate of group 0 to 1.0507e-04.
    Phase: train -->  loss=0.21110609339343178 iou=0.7671038412390204 f1=0.8535166069470213


100%|██████████| 100/100 [01:44<00:00,  1.04s/it, iou=0.775, loss=0.279, f1=0.712]


    Phase: val -->  loss=0.2789513244479895 iou=0.7754021580990411 f1=0.7119949630990934
Not saving for iou 0.7754021580990411

Epoch: 11


100%|██████████| 27/27 [01:21<00:00,  3.01s/it, iou=0.784, loss=0.191, f1=0.877]


Adjusting learning rate of group 0 to 1.0614e-04.
    Phase: train -->  loss=0.19090362113934975 iou=0.7843990776667165 f1=0.8766920081138972


 33%|███▎      | 33/100 [00:36<01:10,  1.06s/it, iou=0.782, loss=0.271, f1=0.718]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
 79%|███████▉  | 79/100 [01:23<00:20,  1.01it/s, iou=0.774, loss=0.283, f1=0.704]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
100%|██████████| 100/100 [01:45<00:00,  1.05s/it, iou=0.78, loss=0.278, f1=0.71]


    Phase: val -->  loss=0.2781154661625624 iou=0.7795251097290712 f1=0.7099721642374172
Not saving for iou 0.7795251097290712

Epoch: 12


100%|██████████| 27/27 [01:20<00:00,  2.97s/it, iou=0.778, loss=0.188, f1=0.867]


Adjusting learning rate of group 0 to 1.0730e-04.
    Phase: train -->  loss=0.18789357664408507 iou=0.7783175101710316 f1=0.8674569088147601


 89%|████████▉ | 89/100 [01:29<00:11,  1.02s/it, iou=0.779, loss=0.278, f1=0.711]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
100%|██████████| 100/100 [01:40<00:00,  1.01s/it, iou=0.779, loss=0.281, f1=0.699]


    Phase: val -->  loss=0.2808697174862027 iou=0.7793595166869569 f1=0.6986922536086686
Not saving for iou 0.7793595166869569

Epoch: 13


100%|██████████| 27/27 [01:19<00:00,  2.95s/it, iou=0.775, loss=0.188, f1=0.867]


Adjusting learning rate of group 0 to 1.0856e-04.
    Phase: train -->  loss=0.1883132501884743 iou=0.7751332661178855 f1=0.866719049875814


100%|██████████| 100/100 [01:42<00:00,  1.02s/it, iou=0.773, loss=0.277, f1=0.68]


    Phase: val -->  loss=0.27677577011287213 iou=0.7726559553028796 f1=0.6803111720180997
Not saving for iou 0.7726559553028796

Epoch: 14


100%|██████████| 27/27 [01:21<00:00,  3.04s/it, iou=0.792, loss=0.17, f1=0.874]


Adjusting learning rate of group 0 to 1.0993e-04.
    Phase: train -->  loss=0.17030952429329907 iou=0.7924526378491756 f1=0.8742572580563843


100%|██████████| 100/100 [01:46<00:00,  1.06s/it, iou=0.786, loss=0.264, f1=0.711]


    Phase: val -->  loss=0.26407353099435565 iou=0.7863215668501556 f1=0.7108753154549164
Saving model with iou 0.7863215668501556

Epoch: 15


100%|██████████| 27/27 [01:22<00:00,  3.05s/it, iou=0.787, loss=0.17, f1=0.87]


Adjusting learning rate of group 0 to 1.1139e-04.
    Phase: train -->  loss=0.17025351524353027 iou=0.7871793114845284 f1=0.8696482233862338


100%|██████████| 100/100 [01:44<00:00,  1.05s/it, iou=0.765, loss=0.282, f1=0.69]


    Phase: val -->  loss=0.28195424389094115 iou=0.7645347780208669 f1=0.6898659808413546
Not saving for iou 0.7645347780208669

Epoch: 16


100%|██████████| 27/27 [01:22<00:00,  3.06s/it, iou=0.773, loss=0.176, f1=0.861]


Adjusting learning rate of group 0 to 1.1295e-04.
    Phase: train -->  loss=0.17631637691347687 iou=0.7729162626205351 f1=0.8613236784009312


100%|██████████| 100/100 [01:41<00:00,  1.01s/it, iou=0.791, loss=0.255, f1=0.72]


    Phase: val -->  loss=0.2553041699156165 iou=0.7908508641950667 f1=0.7199721926874924
Saving model with iou 0.7908508641950667

Epoch: 17


100%|██████████| 27/27 [01:21<00:00,  3.02s/it, iou=0.769, loss=0.173, f1=0.863]


Adjusting learning rate of group 0 to 1.1461e-04.
    Phase: train -->  loss=0.17260597977373335 iou=0.769419377232914 f1=0.8634271790358894


 33%|███▎      | 33/100 [00:34<01:07,  1.00s/it, iou=0.73, loss=0.298, f1=0.627]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
 71%|███████   | 71/100 [01:12<00:28,  1.01it/s, iou=0.754, loss=0.284, f1=0.648]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
100%|██████████| 100/100 [01:42<00:00,  1.03s/it, iou=0.782, loss=0.261, f1=0.689]


    Phase: val -->  loss=0.2612597417086363 iou=0.7815163496548955 f1=0.6888906226814591
Not saving for iou 0.7815163496548955

Epoch: 18


100%|██████████| 27/27 [01:21<00:00,  3.00s/it, iou=0.805, loss=0.146, f1=0.889]


Adjusting learning rate of group 0 to 1.1637e-04.
    Phase: train -->  loss=0.1463832300570276 iou=0.8052454432854359 f1=0.8886420130326875


100%|██████████| 100/100 [01:45<00:00,  1.06s/it, iou=0.787, loss=0.253, f1=0.714]


    Phase: val -->  loss=0.25335914108902213 iou=0.7873046982419742 f1=0.7143001298188334
Not saving for iou 0.7873046982419742

Epoch: 19


100%|██████████| 27/27 [01:21<00:00,  3.01s/it, iou=0.789, loss=0.155, f1=0.878]


Adjusting learning rate of group 0 to 1.1823e-04.
    Phase: train -->  loss=0.15511292467514673 iou=0.789433600871608 f1=0.8784859238512596


 74%|███████▍  | 74/100 [01:18<00:26,  1.03s/it, iou=0.803, loss=0.236, f1=0.729]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
 92%|█████████▏| 92/100 [01:37<00:08,  1.09s/it, iou=0.802, loss=0.238, f1=0.72]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
100%|██████████| 100/100 [01:45<00:00,  1.05s/it, iou=0.797, loss=0.239, f1=0.718]


    Phase: val -->  loss=0.23855223163962364 iou=0.7967314203902099 f1=0.7184085004049714
Saving model with iou 0.7967314203902099

Epoch: 20


100%|██████████| 27/27 [01:21<00:00,  3.01s/it, iou=0.802, loss=0.146, f1=0.884]


Adjusting learning rate of group 0 to 1.2018e-04.
    Phase: train -->  loss=0.14639206285829898 iou=0.8021682678450973 f1=0.8842012824230537


  8%|▊         | 8/100 [00:09<01:43,  1.13s/it, iou=0.462, loss=0.418, f1=0.471]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
 58%|█████▊    | 58/100 [01:01<00:49,  1.19s/it, iou=0.757, loss=0.264, f1=0.67]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
100%|██████████| 100/100 [01:43<00:00,  1.04s/it, iou=0.789, loss=0.251, f1=0.696]


    Phase: val -->  loss=0.2512263326346874 iou=0.7885755562933864 f1=0.6958729168167543
Not saving for iou 0.7885755562933864

Epoch: 21


100%|██████████| 27/27 [01:22<00:00,  3.05s/it, iou=0.784, loss=0.159, f1=0.869]


Adjusting learning rate of group 0 to 1.2223e-04.
    Phase: train -->  loss=0.15934850523869196 iou=0.7839195737705662 f1=0.8689176923489682


100%|██████████| 100/100 [01:45<00:00,  1.06s/it, iou=0.782, loss=0.241, f1=0.745]


    Phase: val -->  loss=0.2411378576606512 iou=0.7819626362156713 f1=0.7445495462428435
Not saving for iou 0.7819626362156713

Epoch: 22


100%|██████████| 27/27 [01:21<00:00,  3.00s/it, iou=0.808, loss=0.138, f1=0.893]


Adjusting learning rate of group 0 to 1.2438e-04.
    Phase: train -->  loss=0.13790391440744754 iou=0.8080343822471018 f1=0.8927291506046786


 98%|█████████▊| 98/100 [01:44<00:02,  1.05s/it, iou=0.783, loss=0.233, f1=0.728]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
100%|██████████| 100/100 [01:46<00:00,  1.07s/it, iou=0.782, loss=0.236, f1=0.72]


    Phase: val -->  loss=0.23614970367401839 iou=0.7820243746801236 f1=0.7202356019184138
Not saving for iou 0.7820243746801236

Epoch: 23


100%|██████████| 27/27 [01:20<00:00,  2.97s/it, iou=0.784, loss=0.153, f1=0.87]


Adjusting learning rate of group 0 to 1.2663e-04.
    Phase: train -->  loss=0.1532798072806111 iou=0.7844053627933809 f1=0.8696097927279484


 56%|█████▌    | 56/100 [00:59<00:47,  1.08s/it, iou=0.769, loss=0.237, f1=0.758]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
 86%|████████▌ | 86/100 [01:30<00:14,  1.05s/it, iou=0.779, loss=0.247, f1=0.713]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
100%|██████████| 100/100 [01:44<00:00,  1.05s/it, iou=0.784, loss=0.246, f1=0.707]


    Phase: val -->  loss=0.24559742964804174 iou=0.783964048130067 f1=0.7066694470711444
Not saving for iou 0.783964048130067

Epoch: 24


100%|██████████| 27/27 [01:22<00:00,  3.06s/it, iou=0.818, loss=0.131, f1=0.895]


Adjusting learning rate of group 0 to 1.2897e-04.
    Phase: train -->  loss=0.13121313022242653 iou=0.8176186858154338 f1=0.8952949975635155


 13%|█▎        | 13/100 [00:14<01:27,  1.00s/it, iou=0.882, loss=0.159, f1=0.841]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
 78%|███████▊  | 78/100 [01:23<00:21,  1.02it/s, iou=0.815, loss=0.23, f1=0.715]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
100%|██████████| 100/100 [01:46<00:00,  1.07s/it, iou=0.796, loss=0.235, f1=0.709]


    Phase: val -->  loss=0.23517467238008977 iou=0.7956961305350236 f1=0.7087413379914085
Not saving for iou 0.7956961305350236

Epoch: 25


100%|██████████| 27/27 [01:22<00:00,  3.05s/it, iou=0.824, loss=0.124, f1=0.899]


Adjusting learning rate of group 0 to 1.3140e-04.
    Phase: train -->  loss=0.12424378924899632 iou=0.8244529081595039 f1=0.8991696499835485


100%|██████████| 100/100 [01:43<00:00,  1.04s/it, iou=0.772, loss=0.251, f1=0.689]


    Phase: val -->  loss=0.25125137813389303 iou=0.7721070517214327 f1=0.6888495339319912
Not saving for iou 0.7721070517214327

Epoch: 26


100%|██████████| 27/27 [01:22<00:00,  3.06s/it, iou=0.818, loss=0.129, f1=0.893]


Adjusting learning rate of group 0 to 1.3393e-04.
    Phase: train -->  loss=0.12856212975802245 iou=0.817708000624313 f1=0.8932071994321181


100%|██████████| 100/100 [01:42<00:00,  1.02s/it, iou=0.795, loss=0.241, f1=0.721]


    Phase: val -->  loss=0.24085068300366402 iou=0.7954293129153308 f1=0.7210413852048645
Not saving for iou 0.7954293129153308

Epoch: 27


100%|██████████| 27/27 [01:20<00:00,  2.98s/it, iou=0.826, loss=0.123, f1=0.901]


Adjusting learning rate of group 0 to 1.3656e-04.
    Phase: train -->  loss=0.12268547299835417 iou=0.8261932614405826 f1=0.9013865226737484


  9%|▉         | 9/100 [00:10<01:47,  1.18s/it, iou=0.87, loss=0.112, f1=0.916]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
 74%|███████▍  | 74/100 [01:17<00:30,  1.16s/it, iou=0.81, loss=0.222, f1=0.749]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
100%|██████████| 100/100 [01:44<00:00,  1.04s/it, iou=0.805, loss=0.235, f1=0.713]


    Phase: val -->  loss=0.2351781737431884 iou=0.8047683930906439 f1=0.713339585719692
Saving model with iou 0.8047683930906439

Epoch: 28


100%|██████████| 27/27 [01:21<00:00,  3.01s/it, iou=0.813, loss=0.133, f1=0.883]


Adjusting learning rate of group 0 to 1.3927e-04.
    Phase: train -->  loss=0.133145356895747 iou=0.8132582499790835 f1=0.8832658898636531


 54%|█████▍    | 54/100 [00:56<00:45,  1.01it/s, iou=0.758, loss=0.224, f1=0.75]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
 75%|███████▌  | 75/100 [01:18<00:27,  1.10s/it, iou=0.763, loss=0.224, f1=0.751]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
100%|██████████| 100/100 [01:45<00:00,  1.05s/it, iou=0.789, loss=0.214, f1=0.76]


    Phase: val -->  loss=0.21442609671503304 iou=0.7894681645381748 f1=0.7604211335345755
Not saving for iou 0.7894681645381748

Epoch: 29


100%|██████████| 27/27 [01:22<00:00,  3.05s/it, iou=0.815, loss=0.13, f1=0.891]


Adjusting learning rate of group 0 to 1.4208e-04.
    Phase: train -->  loss=0.1298415426302839 iou=0.8145775508168069 f1=0.8913893153254996


 22%|██▏       | 22/100 [00:22<01:20,  1.03s/it, iou=0.736, loss=0.289, f1=0.668]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
 82%|████████▏ | 82/100 [01:25<00:18,  1.05s/it, iou=0.791, loss=0.233, f1=0.725]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
100%|██████████| 100/100 [01:44<00:00,  1.04s/it, iou=0.797, loss=0.231, f1=0.722]


    Phase: val -->  loss=0.23114128502085804 iou=0.7967698430719369 f1=0.7220174422700306
Not saving for iou 0.7967698430719369

Epoch: 30


100%|██████████| 27/27 [01:21<00:00,  3.03s/it, iou=0.826, loss=0.122, f1=0.898]


Adjusting learning rate of group 0 to 1.4498e-04.
    Phase: train -->  loss=0.12195124118416398 iou=0.8256461854650858 f1=0.8981404303767645


 10%|█         | 10/100 [00:10<01:38,  1.09s/it, iou=0.885, loss=0.154, f1=0.817]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
 26%|██▌       | 26/100 [00:26<01:23,  1.13s/it, iou=0.787, loss=0.2, f1=0.773]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
 89%|████████▉ | 89/100 [01:34<00:11,  1.08s/it, iou=0.78, loss=0.231, f1=0.716]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined an

    Phase: val -->  loss=0.23326818764209747 iou=0.7757616757176106 f1=0.7194664697006958
Not saving for iou 0.7757616757176106

Epoch: 31


100%|██████████| 27/27 [01:21<00:00,  3.02s/it, iou=0.82, loss=0.125, f1=0.892]


Adjusting learning rate of group 0 to 1.4798e-04.
    Phase: train -->  loss=0.12530078987280527 iou=0.8199538605385367 f1=0.8921893088374737


  2%|▏         | 2/100 [00:02<01:59,  1.22s/it, iou=0.814, loss=0.115, f1=0.899]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
 49%|████▉     | 49/100 [00:52<00:50,  1.02it/s, iou=0.801, loss=0.223, f1=0.728]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
 80%|████████  | 80/100 [01:26<00:22,  1.14s/it, iou=0.804, loss=0.216, f1=0.735]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined 

    Phase: val -->  loss=0.2341567036882043 iou=0.7884829035836384 f1=0.7107798516759852
Not saving for iou 0.7884829035836384

Epoch: 32


100%|██████████| 27/27 [01:21<00:00,  3.03s/it, iou=0.803, loss=0.141, f1=0.878]


Adjusting learning rate of group 0 to 1.5106e-04.
    Phase: train -->  loss=0.1407895595939071 iou=0.8029528390156313 f1=0.8775041620659546


100%|██████████| 100/100 [01:43<00:00,  1.03s/it, iou=0.799, loss=0.216, f1=0.75]


    Phase: val -->  loss=0.2162415268458426 iou=0.798832008316442 f1=0.7498531988713699
Not saving for iou 0.798832008316442

Epoch: 33


100%|██████████| 27/27 [01:22<00:00,  3.06s/it, iou=0.827, loss=0.117, f1=0.901]


Adjusting learning rate of group 0 to 1.5424e-04.
    Phase: train -->  loss=0.11744792676634258 iou=0.8267209762547383 f1=0.9009288841333553


100%|██████████| 100/100 [01:46<00:00,  1.07s/it, iou=0.797, loss=0.215, f1=0.745]


    Phase: val -->  loss=0.2146270587295294 iou=0.7970767819168187 f1=0.7450346357360396
Not saving for iou 0.7970767819168187

Epoch: 34


100%|██████████| 27/27 [01:22<00:00,  3.05s/it, iou=0.811, loss=0.127, f1=0.893]


Adjusting learning rate of group 0 to 1.5750e-04.
    Phase: train -->  loss=0.12749935548614572 iou=0.8109477065008349 f1=0.8927811460588235


 59%|█████▉    | 59/100 [01:01<00:44,  1.08s/it, iou=0.806, loss=0.225, f1=0.703]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
 66%|██████▌   | 66/100 [01:08<00:37,  1.10s/it, iou=0.806, loss=0.229, f1=0.691]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
 87%|████████▋ | 87/100 [01:29<00:12,  1.03it/s, iou=0.809, loss=0.219, f1=0.707]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined

    Phase: val -->  loss=0.2272319607809186 iou=0.7846528315004245 f1=0.7078016874976923
Not saving for iou 0.7846528315004245

Epoch: 35


100%|██████████| 27/27 [01:21<00:00,  3.01s/it, iou=0.834, loss=0.114, f1=0.904]


Adjusting learning rate of group 0 to 1.6085e-04.
    Phase: train -->  loss=0.11436805846514525 iou=0.833666998852881 f1=0.9044258663590495


 39%|███▉      | 39/100 [00:42<01:05,  1.07s/it, iou=0.785, loss=0.227, f1=0.751]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
100%|██████████| 100/100 [01:44<00:00,  1.04s/it, iou=0.797, loss=0.218, f1=0.757]


    Phase: val -->  loss=0.2175805270858109 iou=0.7970939538471525 f1=0.7574296294377402
Not saving for iou 0.7970939538471525

Epoch: 36


100%|██████████| 27/27 [01:20<00:00,  3.00s/it, iou=0.832, loss=0.113, f1=0.906]


Adjusting learning rate of group 0 to 1.6429e-04.
    Phase: train -->  loss=0.11315703005702407 iou=0.8317338715144479 f1=0.905740875707456


100%|██████████| 100/100 [01:45<00:00,  1.06s/it, iou=0.795, loss=0.223, f1=0.731]


    Phase: val -->  loss=0.22265656900592148 iou=0.7953747734524063 f1=0.7310981116406988
Not saving for iou 0.7953747734524063

Epoch: 37


100%|██████████| 27/27 [01:21<00:00,  3.02s/it, iou=0.842, loss=0.108, f1=0.906]


Adjusting learning rate of group 0 to 1.6782e-04.
    Phase: train -->  loss=0.10819393630932879 iou=0.8416153082776066 f1=0.9061675845976324


 36%|███▌      | 36/100 [00:37<01:01,  1.05it/s, iou=0.784, loss=0.243, f1=0.707]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
 39%|███▉      | 39/100 [00:40<01:06,  1.09s/it, iou=0.796, loss=0.243, f1=0.701]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
100%|██████████| 100/100 [01:44<00:00,  1.04s/it, iou=0.814, loss=0.216, f1=0.743]


    Phase: val -->  loss=0.21603019520640374 iou=0.8143358869038733 f1=0.7431837448744342
Saving model with iou 0.8143358869038733

Epoch: 38


100%|██████████| 27/27 [01:21<00:00,  3.03s/it, iou=0.843, loss=0.107, f1=0.912]


Adjusting learning rate of group 0 to 1.7144e-04.
    Phase: train -->  loss=0.10737562897028746 iou=0.8428500448325422 f1=0.9120952802731381


 12%|█▏        | 12/100 [00:12<01:25,  1.03it/s, iou=0.661, loss=0.318, f1=0.663]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
 83%|████████▎ | 83/100 [01:26<00:17,  1.04s/it, iou=0.803, loss=0.224, f1=0.724]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
100%|██████████| 100/100 [01:43<00:00,  1.03s/it, iou=0.796, loss=0.228, f1=0.719]


    Phase: val -->  loss=0.22769822172820567 iou=0.7960235070618114 f1=0.7188858075528388
Not saving for iou 0.7960235070618114

Epoch: 39


100%|██████████| 27/27 [01:20<00:00,  3.00s/it, iou=0.833, loss=0.111, f1=0.906]


Adjusting learning rate of group 0 to 1.7514e-04.
    Phase: train -->  loss=0.11056645435315592 iou=0.8329538741746768 f1=0.9062623641583106


 49%|████▉     | 49/100 [00:51<00:56,  1.12s/it, iou=0.818, loss=0.202, f1=0.746]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
 62%|██████▏   | 62/100 [01:04<00:41,  1.09s/it, iou=0.804, loss=0.208, f1=0.734]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
 69%|██████▉   | 69/100 [01:11<00:31,  1.02s/it, iou=0.8, loss=0.213, f1=0.73]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined an

    Phase: val -->  loss=0.2020579175837338 iou=0.8133752303424105 f1=0.7501151450738912
Not saving for iou 0.8133752303424105

Epoch: 40


100%|██████████| 27/27 [01:20<00:00,  2.97s/it, iou=0.835, loss=0.11, f1=0.907]


Adjusting learning rate of group 0 to 1.7892e-04.
    Phase: train -->  loss=0.1098669639064206 iou=0.835428675797295 f1=0.9073496411972762


  8%|▊         | 8/100 [00:09<01:36,  1.04s/it, iou=0.887, loss=0.0866, f1=0.934]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
 14%|█▍        | 14/100 [00:15<01:33,  1.09s/it, iou=0.866, loss=0.14, f1=0.851]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
 89%|████████▉ | 89/100 [01:34<00:11,  1.01s/it, iou=0.812, loss=0.217, f1=0.728]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined 

    Phase: val -->  loss=0.22387346442788839 iou=0.8109673127803694 f1=0.7119409435641529
Not saving for iou 0.8109673127803694

Epoch: 41


100%|██████████| 27/27 [01:21<00:00,  3.01s/it, iou=0.841, loss=0.105, f1=0.908]


Adjusting learning rate of group 0 to 1.8279e-04.
    Phase: train -->  loss=0.10452385577890608 iou=0.8406662566170362 f1=0.9084133061326748


100%|██████████| 100/100 [01:45<00:00,  1.05s/it, iou=0.804, loss=0.224, f1=0.711]


    Phase: val -->  loss=0.22428470924496652 iou=0.8044263453111457 f1=0.7114169794398104
Not saving for iou 0.8044263453111457

Epoch: 42


100%|██████████| 27/27 [01:21<00:00,  3.02s/it, iou=0.83, loss=0.112, f1=0.902]


Adjusting learning rate of group 0 to 1.8674e-04.
    Phase: train -->  loss=0.11158533435728815 iou=0.8298196577092389 f1=0.9018438483403364


100%|██████████| 100/100 [01:45<00:00,  1.05s/it, iou=0.798, loss=0.209, f1=0.76]


    Phase: val -->  loss=0.20874200876802207 iou=0.7978399270019332 f1=0.7597377519942089
Not saving for iou 0.7978399270019332

Epoch: 43


100%|██████████| 27/27 [01:21<00:00,  3.01s/it, iou=0.832, loss=0.118, f1=0.897]


Adjusting learning rate of group 0 to 1.9077e-04.
    Phase: train -->  loss=0.11836730316281319 iou=0.8316060376063121 f1=0.8972417575635667


 33%|███▎      | 33/100 [00:34<01:13,  1.10s/it, iou=0.827, loss=0.219, f1=0.719]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
100%|██████████| 100/100 [01:44<00:00,  1.05s/it, iou=0.797, loss=0.215, f1=0.73]


    Phase: val -->  loss=0.2147300520911813 iou=0.7968554166403501 f1=0.7304019959734984
Not saving for iou 0.7968554166403501

Epoch: 44


100%|██████████| 27/27 [01:22<00:00,  3.05s/it, iou=0.841, loss=0.105, f1=0.909]


Adjusting learning rate of group 0 to 1.9489e-04.
    Phase: train -->  loss=0.10481190515889062 iou=0.8405121100495352 f1=0.9092509353670833


100%|██████████| 100/100 [01:46<00:00,  1.06s/it, iou=0.804, loss=0.2, f1=0.766]


    Phase: val -->  loss=0.19990453304722905 iou=0.8038900503137097 f1=0.7662288986189085
Not saving for iou 0.8038900503137097

Epoch: 45


100%|██████████| 27/27 [01:22<00:00,  3.07s/it, iou=0.839, loss=0.111, f1=0.904]


Adjusting learning rate of group 0 to 1.9908e-04.
    Phase: train -->  loss=0.11089416648502703 iou=0.8389553627435163 f1=0.9041632228444314


 71%|███████   | 71/100 [01:14<00:30,  1.07s/it, iou=0.811, loss=0.195, f1=0.751]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
100%|██████████| 100/100 [01:43<00:00,  1.03s/it, iou=0.808, loss=0.198, f1=0.754]


    Phase: val -->  loss=0.19847206639125944 iou=0.8083259615844757 f1=0.7541060471916117
Not saving for iou 0.8083259615844757

Epoch: 46


100%|██████████| 27/27 [01:22<00:00,  3.04s/it, iou=0.843, loss=0.105, f1=0.908]


Adjusting learning rate of group 0 to 2.0336e-04.
    Phase: train -->  loss=0.10456694193460324 iou=0.8427090023682933 f1=0.9076106337978225


 17%|█▋        | 17/100 [00:18<01:33,  1.13s/it, iou=0.81, loss=0.202, f1=0.765]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
 57%|█████▋    | 57/100 [00:59<00:44,  1.03s/it, iou=0.827, loss=0.207, f1=0.737]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
 61%|██████    | 61/100 [01:03<00:35,  1.10it/s, iou=0.828, loss=0.209, f1=0.728]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined 

    Phase: val -->  loss=0.2195779182948172 iou=0.8017505098208486 f1=0.7218455592509433
Not saving for iou 0.8017505098208486

Epoch: 47


100%|██████████| 27/27 [01:21<00:00,  3.01s/it, iou=0.827, loss=0.114, f1=0.899]


Adjusting learning rate of group 0 to 2.0771e-04.
    Phase: train -->  loss=0.11371588030899013 iou=0.8274347712511622 f1=0.899024709912653


 37%|███▋      | 37/100 [00:38<00:59,  1.07it/s, iou=0.8, loss=0.204, f1=0.785]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
100%|██████████| 100/100 [01:43<00:00,  1.04s/it, iou=0.769, loss=0.236, f1=0.724]


    Phase: val -->  loss=0.23606291491538287 iou=0.7690888928256001 f1=0.723749054988133
Not saving for iou 0.7690888928256001

Epoch: 48


100%|██████████| 27/27 [01:22<00:00,  3.04s/it, iou=0.845, loss=0.103, f1=0.913]


Adjusting learning rate of group 0 to 2.1214e-04.
    Phase: train -->  loss=0.10287746124797398 iou=0.8453574878969555 f1=0.9133672865552805


 61%|██████    | 61/100 [01:04<00:44,  1.13s/it, iou=0.792, loss=0.209, f1=0.744]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
100%|██████████| 100/100 [01:46<00:00,  1.07s/it, iou=0.798, loss=0.225, f1=0.717]


    Phase: val -->  loss=0.224982679951936 iou=0.798317805448603 f1=0.7166415217452827
Not saving for iou 0.798317805448603

Epoch: 49


100%|██████████| 27/27 [01:22<00:00,  3.06s/it, iou=0.852, loss=0.095, f1=0.92]


Adjusting learning rate of group 0 to 2.1664e-04.
    Phase: train -->  loss=0.0950218346659784 iou=0.8519546896107364 f1=0.9197298142310412


 78%|███████▊  | 78/100 [01:21<00:20,  1.07it/s, iou=0.765, loss=0.231, f1=0.717]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
100%|██████████| 100/100 [01:43<00:00,  1.04s/it, iou=0.773, loss=0.236, f1=0.703]


    Phase: val -->  loss=0.23618650045245887 iou=0.773351807572374 f1=0.7034582489674588
Not saving for iou 0.773351807572374

Epoch: 50


100%|██████████| 27/27 [01:20<00:00,  2.99s/it, iou=0.826, loss=0.114, f1=0.896]


Adjusting learning rate of group 0 to 2.2122e-04.
    Phase: train -->  loss=0.11352154364188512 iou=0.826094043302827 f1=0.8956781544971301


 39%|███▉      | 39/100 [00:42<01:09,  1.13s/it, iou=0.764, loss=0.231, f1=0.732]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
 58%|█████▊    | 58/100 [01:01<00:43,  1.04s/it, iou=0.779, loss=0.224, f1=0.727]/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1609: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 due to no true nor predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, "true nor predicted", "F-score is", len(true_sum))
100%|██████████| 100/100 [01:45<00:00,  1.05s/it, iou=0.786, loss=0.215, f1=0.742]

    Phase: val -->  loss=0.21549396131187679 iou=0.7860229577679266 f1=0.7423885815566308
Not saving for iou 0.7860229577679266


### Results
normal + kleine lr

loss=0.27545809376984837 iou=0.7928296608501525 f1=0.7245020875022741


---
Crop + 50Epoch

loss=0.21603019520640374 iou=0.8143358869038733 f1=0.7431837448744342
